builtin/logical/database/backend.go

```diff
@@ -228,6 +272,156 @@ func (b *databaseBackend) GetConnection(ctx context.Context, s logical.Storage,
 	return db, nil
 }
 
+// initQueue preforms the necessary checks and initializations needed to preform
+// automatic credential rotation for roles associated with static accounts. This
+// method verifies if a queue is needed (primary server or local mount), and if
+// so initializes the queue and launches a go-routine to periodically invoke a
+// method to preform the rotations.
+//
+// initQueue is invoked by the Factory method in a go-routine. The Factory does
+// not wait for success or failure of it's tasks before continuing. This is to
+// avoid blocking the mount process while loading and evaluating existing roles,
+// etc.
+func (b *databaseBackend) initQueue(ctx context.Context, conf *logical.BackendConfig) {
+	// verify this mount is on the primary server, or is a local mount. If not, do
+	// not create a queue or launch a ticker. Both processing the WAL list and
+	// populating the queue are done sequentially and before launching a
+	// go-routine to run the periodic ticker.
+	replicationState := conf.System.ReplicationState()
+	if (conf.System.LocalMount() || !replicationState.HasState(consts.ReplicationPerformanceSecondary)) &&
+		!replicationState.HasState(consts.ReplicationDRSecondary) &&
+		!replicationState.HasState(consts.ReplicationPerformanceStandby) {
+		b.Logger().Info("initializing database rotation queue")
+
+		// Sleep a few seconds to allow Vault to mount and complete setup, so
+		// that we can write to storage
+		time.Sleep(3 * time.Second)
```

Comment:
Is there a better way to do this?  I just wonder if we can have a more explicit test.

Reply:
I would also love to see this codified in the system backend.  Basically a function that says should I be running things.

Reply:
I agree! I ran into an issue in the `loadStaticWALs` method where I was calling `framework.DeleteWAL` and getting errors because the storage couldn't be written to during startup. 

Perhaps adding and exposing an `IsReadOnly()` or `IsSetup()` method to `logical.Storage` that uses connects down to `BarrierView` that uses checks method?

- https://github.com/hashicorp/vault/blob/e4977bb257ee8ff0bbe4c30c118169dd8f4856e9/vault/barrier_view.go#L43

I believe the [ErrSetupReadOnly](https://github.com/hashicorp/vault/blob/cdcr-backend-grpc/sdk/logical/storage.go#L20) is being triggered in that file. I'm not sure how or where to bubble that up to the  `logical.Storage` but I could find out

Reply:
Here is some prior art from the kv plugin: https://github.com/hashicorp/vault-plugin-secrets-kv/blob/master/upgrade.go#L198-L213

Adding a helper to do this is a good idea. Not sure it's a requirement for this PR though.

Reply:
I added a task to our internal tracker to pick this up as a follow-up after this PR, sound good?